lib/src/Uncertainty/Process/ConditionedGaussianProcess.cxx

```diff
@@ -98,7 +98,47 @@ void ConditionedGaussianProcess::initialize()
   CovarianceMatrix covarianceMatrix(krigingResult_.getConditionalCovariance(vertices));
   // Get the Cholesky factor
   LOGINFO(OSS(false) << "Evaluation of the Cholesky factor");
-  covarianceCholeskyFactor_ = covarianceMatrix.computeCholesky();
+
+  // Boolean flag to tell if the regularization is enough
+  Bool continuationCondition = true;
+  // Scaling factor of the matrix : M-> M + \lambda I with \lambda very small
+  // The regularization is needed for fast decreasing covariance models
+  Scalar maxEV = -1.0;
+  const Scalar startingScaling = ResourceMap::GetAsScalar("GaussianProcess-StartingScaling");
+  const Scalar maximalScaling = ResourceMap::GetAsScalar("GaussianProcess-MaximalScaling");
+  Scalar cumulatedScaling = 0.0;
+  Scalar scaling = startingScaling;
+  while (continuationCondition)
+  {
+    try
+    {
+      covarianceCholeskyFactor_ = covarianceMatrix.computeCholesky();
+      continuationCondition = false;
+    }
+    // If the factorization failed regularize the matrix
+    // Here we use a generic exception as different exceptions may be thrown
+    catch (Exception &)
+    {
+      // If the largest eigenvalue module has not been computed yet...
+      if (maxEV < 0.0)
+      {
+        maxEV = covarianceMatrix.computeLargestEigenValueModule();
```

Comment:
I am not sure the maximum eigenvalue should be the reference quantify used for to scale lambda. The maximum eigenvalue of the *correlation* matrix necessarily belongs to [1, n], where n is the number of columns in the matrix (while all other eigenvalues can be arbitrarily small). The largest eigenvalue of the *covariance* matrix is also unlikely IMO to give much insight into the proper value of lambda. When the conditioning of the covariance matrix is poor, it is rather because the smallest eigenvalue is too small than because the largest eigenvalue is too large.

Theoretically, I think the criterion should rather be built around the smallest eigenvalue. For example, we could start with lambda equal to the smallest eigenvalue and stop the algorithm when it reaches some multiple of the smallest eigenvalue.

Failing that, we could base the criterion on the determinant of the covariance matrix (det^{1/n} for example) : it is still more informative than the largest eigenvalue.

Regardless, your proposal is a clear improvement so I am in favor of it being merged.

Reply:
If a conditioning point is used, then I think it should be excluded from the computation of the covariance matrix: after all, the value of the conditioned process at observation points is deterministic.

Reply:
@josephmure The use off the maximum eigenvalue as a reference scale is to avoid cancellation of the regulartization during the floating point operations. The condition number is a ratio between the largest and the smallest eigenvalues, so it cannot be reduced to any of these eigenvalues. The way I perform the regularization is to ensure that the matrix is perturbed as slightly as possible in the most important direction, the one associated to the largest eigenvalue, but with an actual perturbation also in this direction.

Your idea of excluding training points from the sampling sounds good to me. I will see if I can implement it on time.

Reply:
In any case, a regularization loop is a good thing as the covariance matrix may be ill conditioned even if no training point is present in the mesh vertices...

Reply:
I started a quite complex implementation before to realize that it was trivial to detect if training points were among the mesh vertices. BTW a geometrical check would not have been enough as training points can be in the set of vertices without creating degenerated covariance matrices if the nugget factor is positive. Here are the main comments in ConditionedGaussianProcess::initialize():
```  // Now check if there is any point both in the input sample and in the mesh vertices.
  // They are characterized by
  // a zero cross: a null row and a null column which cross at a zero diagonal element.
  // The trick is to replace this value by the maximum marginal variance, then to
  // remember to set to zero the value at the corresponding index during the sampling phase
```
and:
```
        // Then put the maximum variance on the diagonal.
        // In theory any positive number should work but
        // this way the condition number should be greatly improved.
```